plugins/package-managers/node/src/main/kotlin/utils/NpmDetection.kt

```diff
@@ -17,237 +17,223 @@
+     * return null otherwise.
+     */
+    open fun getWorkspaces(projectDir: File): List<String>? {
+        val workspaceFile = projectDir.resolve(workspaceFileName)
+        if (!workspaceFile.isFile) return null
+
+        val workspaces = runCatching {
+            workspaceFile.readTree().get("workspaces")
+        }.onFailure {
+            logger.error { "Failed to parse '$workspaceFile': ${it.collectMessages()}" }
+        }.getOrNull() ?: return null
+
+        val packages = when {
+            workspaces.isArray -> workspaces
+            workspaces.isObject -> workspaces["packages"]
+            else -> null
+        } ?: run {
+            logger.warn { "Unable to read workspaces from '$workspaceFile'." }
+            return null
         }
 
-        return workspaces?.map {
-            val pattern = "glob:${definitionFile.parentFile.invariantSeparatorsPath}/${it.textValue()}"
-            FileSystems.getDefault().getPathMatcher(pattern)
-        }.orEmpty()
-    }
+        return packages.map {
+            val pattern = "${workspaceFile.parentFile.invariantSeparatorsPath}/${it.textValue()}"
 
-    val result = mutableSetOf<File>()
-
-    definitionFiles.forEach { definitionFile ->
-        val workspaceMatchers = getWorkspaceMatchers(definitionFile)
-        workspaceMatchers.forEach { matcher ->
-            definitionFiles.forEach inner@{ other ->
-                // Since yarn workspaces matchers support '*' and '**' to match multiple directories the matcher
-                // cannot be used as is for matching the 'package.json' file. Thus matching against the project
-                // directory since this works out of the box. See also:
-                //   https://github.com/yarnpkg/yarn/issues/3986
-                //   https://github.com/yarnpkg/yarn/pull/5607
-                val projectDir = other.parentFile.toPath()
-                if (other != definitionFile && matcher.matches(projectDir)) {
-                    result += other
-                    return@inner
-                }
-            }
+            // NPM and Yarn treat "*" as an alias for "**", so replace any single "*" with "**".
+            pattern.replace(WORKSPACES_SINGLE_ASTERISK_REGEX, "**")
         }
     }
 
-    return result
+    /**
+     * Return a score for the [projectDir] based on the presence of files specific to this package manager.
+     * The higher the score, the more likely it is that the [projectDir] is managed by this package manager.
+     */
+    fun getFileScore(projectDir: File): Int =
+        listOf(
+            hasLockFile(projectDir),
+            hasFile(projectDir, otherFileName),
+            // Only count the presence of an additional workspace file if it is not the definition file.
+            workspaceFileName != DEFINITION_FILE && hasFile(projectDir, workspaceFileName)
+        ).count { it }
 }
+
+/**
+ * Return true if the [projectDir] contains a non-empty named [file], or return false otherwise.
```

Comment:
"Return true if [file(Name)] is not `null` and the [projectDir] contains a non-empty **file** named [file(Name)]."